### ISSUE TITLE
Fast C io.ascii reader failures with Python 3.8b1

### ISSUE THREAD

**Issue by astrofrog** [astropy/astropy#8804]
I tried running the astropy tests with Python 3.8 and the latest developer version of Cython and Numpy, and there are a number of failures in the io.ascii reader. I started trying to debug this and found that in Python 3.7, ``CParser.__reduce__`` isn't used, and it looks like it is now with 3.8, and there was an inconsistency in the APIs. I put fixes for those in a branch in my repo:

https://github.com/astropy/astropy/compare/master...astrofrog:fix-io-ascii-38

However, I then still see errors such as:

```
Python(38627,0x10d7395c0) malloc: *** error for object 0x100000000: pointer being freed was not allocated
Python(38627,0x10d7395c0) malloc: *** set a breakpoint in malloc_error_break to debug
Python(38626,0x1059865c0) malloc: *** error for object 0x100000000: pointer being freed was not allocated
Python(38626,0x1059865c0) malloc: *** set a breakpoint in malloc_error_break to debug
Python(38625,0x1145535c0) malloc: *** error for object 0x10100000100: pointer being freed was not allocated
Python(38625,0x1145535c0) malloc: *** set a breakpoint in malloc_error_break to debug
Python(38637,0x10dc8e5c0) malloc: *** error for object 0x10100000100: pointer being freed was not allocated
Python(38637,0x10dc8e5c0) malloc: *** set a breakpoint in malloc_error_break to debug
```

Sometimes the tests pass anyway so be sure to run the tests with the ``-s`` flags. The tests run much slower than before, and this is due to a big slow down in the parallel version. I then also run into test failures that are very difficult for me to track down, so I'm giving up for now.

If anyone is interested in trying to get this to work, feel free to take a look at the initial commits in my branch. Another thing that might be useful would be to bisect CPython to see what introduced the change, which might help understand how to fix it (or if it's a CPython bug).

So in short, if you're bored and are familiar with debugging C code, this issue is all yours! 😆

**Comment by astrofrog** [comment]
@dhomeier @saimn - it looks like you both made some fixes to the fast reader in the last couple of years, so pinging you in case you'd be willing to take a look at this!

**Comment by astrofrog** [comment]
Just for info I only see this on MacOS X, I can't reproduce the issue on Linux.

**Comment by astrofrog** [comment]
Ok so I bisected CPython and I found the culprit - a change in the default multiprocessing method used on MacOS X for Python 3.8. The bug actually exists on all Python versions though, it's only apparent in 3.8 because of the change in default. I've opened a separate issue as a result, describing how to reproduce this issue on all Python versions: https://github.com/astropy/astropy/issues/8851.

**Comment by astrofrog** [comment]
I'll close this since #8851 can be used to track this now (since it's not really a bug specific to Python 3.8)
